uber/automated_emails.py

```diff
@@ -1,75 +1,309 @@
+            return False
+
+        for date_filter in self.when:
+            if not date_filter():
+                return False
+
+        return True
 
     def __repr__(self):
         return '<{}: {!r}>'.format(self.__class__.__name__, self.subject)
 
-    def prev(self, x, all_sent=None):
-        if all_sent:
-            return (x.__class__.__name__, x.id, self.ident) in all_sent
-        else:
-            with Session() as session:
-                return session.query(Email).filter_by(model=x.__class__.__name__, fk_id=x.id, ident=self.ident).first()
+    def _already_sent(self, session, model_inst, previously_sent_emails=None):
+        """
+        Returns true if we have a record of previously sending this email
 
-    def should_send(self, x, all_sent=None):
+        CPU Optimization: when using this function as part of batch processing, you can pass in a list of all
+        previously sent emails, previously_sent_emails, in order to avoid having to query the DB for this specific email
+        """
+        if previously_sent_emails is not None:
+            # optimized version: use the cached previously_sent_emails so we don't have to query the DB
+            return (model_inst.__class__.__name__, model_inst.id, self.ident) in previously_sent_emails
+        else:
+            # non-optimized version: query the DB to find any emails that were previously sent from this email category
+            return session.query(Email).filter_by(
+                model=model_inst.__class__.__name__, fk_id=model_inst.id, ident=self.ident).first()
+
+    def attempt_to_send(self, session, model_inst,
+                        approved_idents=None, previously_sent_emails=None, raise_errors=False, daemon_job_results=None):
+        """
+        If it's OK to send an email of our category to this model instance (i.e. a particular Attendee) then send it.
+
+        :param session - current database session
+        :param model_inst - the model (i.e. Attendee or similar) we are checking to see if we should send an email from
+                            our category to this particular model
+        :param approved_idents - OPTIMIZATION: CACHED LIST: a list of approved email category idents
+        :param previously_sent_emails - OPTIMIZATION: CACHED LIST: a list of specific emails which were previously sent
+        :param raise_errors: if False, exceptions will be logged and swallowed here. Useful when running large batches
+        :param daemon_job_results:  if set, we're running as part of the automated email daemon and can use this object
+                                to log information about this sending attempt.
+        """
+        if self._should_send(session, model_inst, approved_idents, previously_sent_emails, daemon_job_results=None):
+            self.send(model_inst, raise_errors=raise_errors)
+
+    def _should_send(self, session, model_inst,
+                     approved_idents=None, previously_sent_emails=None,
```

Comment:
Gonna wait until @binary1230 gives the go-ahead that this is ready for the final review before I give this a thorough overview, but one suggestion here is that it's a bit clunky to make ``approved_idents`` a parameter here.

Instead, now that we have ``@request_cached_property`` we could have a ``c.APPROVED_EMAIL_IDENTS`` property, and then that would get persisted across all function calls.  This would let us go back to just having a ``.approved`` property without requiring a separate database query every time.

The only complication to this approach is that at the beginning of ``send_all`` we'll need to say ``threadlocal.set('uber.config.BADGES_SOLD', None)`` so as to reset the property.  (``@request_cached_property`` resets on every HTTP, JSON-RPC, and WebSocket request, but doesn't reset automatically on ``DaemonTask`` invocations.)